File: Ocis.Server/Ocis.Protocol/ProtocolSpec.rs
pub mod Ocis {
    use super::*;
    pub mod Server {
        use super::*;
        pub mod ProtocolSpec {
            use super::*;
            use fable_library_rust::NativeArray_::Array;
            use fable_library_rust::String_::string;
            /// Request packet
            #[derive(Clone, Debug, Hash, PartialEq, PartialOrd,)]
            pub struct RequestPacket {
                pub MagicNumber: u32,
                pub Version: u8,
                pub CommandType: i32,
                pub TotalPacketLength: i32,
                pub KeyLength: i32,
                pub ValueLength: i32,
                pub Key: Array<u8>,
                pub Value: Option<Array<u8>>,
            }
            impl core::fmt::Display for RequestPacket {
                fn fmt(&self, f: &mut core::fmt::Formatter)
                 -> core::fmt::Result {
                    write!(f, "{}", core::any::type_name::<Self>())
                }
            }
            /// Response packetnse packet
            #[derive(Clone, Debug, Hash, PartialEq, PartialOrd,)]
            pub struct ResponsePacket {
                pub MagicNumber: u32,
                pub Version: u8,
                pub StatusCode: u8,
                pub TotalPacketLength: i32,
                pub ValueLength: i32,
                pub ErrorMessageLength: i32,
                pub Value: Option<Array<u8>>,
                pub ErrorMessage: Option<string>,
            }
            impl core::fmt::Display for ResponsePacket {
                fn fmt(&self, f: &mut core::fmt::Formatter)
                 -> core::fmt::Result {
                    write!(f, "{}", core::any::type_name::<Self>())
                }
            }
        }
    }
}

File: Ocis.Server/ProtocolSpec.fs
module Ocis.Server.ProtocolSpec

/// Protocol magic number "OCIS"
[<Literal>]
let MAGIC_NUMBER = 0x5349434Fu

/// Protocol version
[<Literal>]
let PROTOCOL_VERSION = 1uy

/// Protocol header size (bytes)
[<Literal>]
let HEADER_SIZE = 18

/// Command type
type CommandType =
    | Set = 1
    | Get = 2
    | Delete = 3

/// Status code
type StatusCode =
    | Success = 0uy
    | NotFound = 1uy
    | Error = 2uy

/// Request packet
type RequestPacket =
    { MagicNumber: uint32
      Version: byte
      CommandType: CommandType
      TotalPacketLength: int32
      KeyLength: int32
      ValueLength: int32
      Key: byte array
      Value: byte array option }

/// Response packetnse packet
type ResponsePacket =
    { MagicNumber: uint32
      Version: byte
      StatusCode: StatusCode
      TotalPacketLength: int32
      ValueLength: int32
      ErrorMessageLength: int32
      Value: byte array option
      ErrorMessage: string option }
